### PR TITLE
Implement ICE Restarts

### LINKED ISSUE #1251: No easy way to restart ICE
At peerconnection.c:406, it says:
```
	case options != nil:
		return SessionDescription{}, fmt.Errorf("TODO handle options")
```
I don't see any other way to restart ICE on an existing PC.  (I'm currently working around the issue by tearing down the connection and creating a new one, but that has other problems, such as requiring a keyframe even if the interruption was short-lived.)

### PR DESCRIPTION
**Pull request by Sean-Der**
Resolves #1251 

@jech I added a dedicated demo page to show off the behavior, hopefully that is everything you need!

### CODE REVIEW

**Comment by Sean-Der** [comment]
@at-wat We are good to go! Sorry for all the churn lately, but really wanted to get this out for https://www.irif.fr/~jch/software/sfu/

I will work on fixing tests, bugs and docs to make up for everything you have done the past week :)

**Comment by jeremija** [comment]
This is cool! I just tested the example and it works fine in Chrome, but I've been getting some weird errors in Firefox. It connected the first time, but the restart button did nothing, and now it won't connect :/

**Comment by Sean-Der** [comment]
@jeremija oh mind commented out the `dc.open` handler in the index.html? I think I am just using APIs that haven't been implemented yet!

I think server side it *should* be all good though.

**Comment by jeremija** [comment]
@Sean-Der yeah commenting that out gets rid of the errors in Safari.

Re Firefox something really weird is going in:

I'm getting:

```
UDP socket error:Internal error at /builds/worker/checkouts/gecko/dom/network/UDPSocketParent.cpp:248 this=0x111dd8400

ICE(PC:1593495166890761 (id=6442450945 url=http://localhost:8080/)): failed to find default addresses

getifaddrs error e = 1
```

in `about:webrtc-internals` and in the console:

```
WebRTC: ICE failed, add a TURN server and see about:webrtc for more details
```

I've seen this happen before occasionally, so perhaps it's not a bug in your code.

**Comment by jeremija** [comment]
Update, so if I use `127.0.0.1:8080` Firefox connects, but not if I use `localhost:8080`

File: examples/ice-restart/index.html [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+          </ul>
+        </div>`
+
+        document.getElementById('iceSelectedPairs').appendChild(el.content.firstChild);
+      }, 3000);
+    }
+
+    dc.onmessage = () => {
+      let el = document.createElement('p')
+      el.appendChild(document.createTextNode(event.data))
```
Comment (jeremija) [inline examples/ice-restart/index.html:42]: Is the `event` supposed to be implicitly defined here? This is new for me, but if so, this works in Chrome, but fails in Firefox with:

```
TypeError: event is undefined
```

I had to modify line 40 to `dc.onmessage = (event ) => {`.

Comment (Sean-Der) [reply examples/ice-restart/index.html:42]: Oh wild, didn't realize that implicitly existed O_O

Comment (Sean-Der) [reply examples/ice-restart/index.html:42]: Fixed!

Comment (jeremija) [reply examples/ice-restart/index.html:42]: Gotta love JavaScript ¯\_(ツ)_/¯

**Comment by Sean-Der** [comment]
@jeremija Yea I *think* that was from my earlier run. I did write tests to cover all the good cases, but I should probably write some more error ones.

I wasn't able to get restart working in WASM unfortunately. Not my idea of a fun Monday night though debugging C++ and nodejs though <_<

**Comment by jech** [comment]
I've tried to use this in the Unnamed SFU, and it crashes while attempting to send RTCP during ICE restart.  Perhaps I'm doing something wrong.
```
 github.com/pion/ice/v2.(*candidateBase).writeTo(0xc0000966e0, 0xc00014a3c0, 0x5e, 0xa0, 0xaeac20, 0xc000096790, 0x0, 0x0, 0x0)
         /home/jch/go/pkg/mod/github.com/pion/ice/v2@v2.0.0-rc.6/candidate_base.go:203 +0x4e
 github.com/pion/ice/v2.(*candidatePair).Write(...)
         /home/jch/go/pkg/mod/github.com/pion/ice/v2@v2.0.0-rc.6/candidatepair.go:86
 github.com/pion/ice/v2.(*Conn).Write(0xc00044b120, 0xc00014a3c0, 0x5e, 0xa0, 0xc00014a3c0, 0x5e, 0xa0)
         /home/jch/go/pkg/mod/github.com/pion/ice/v2@v2.0.0-rc.6/transport.go:105 +0x149
 github.com/pion/webrtc/v3/internal/mux.(*Endpoint).Write(0xc00018b5a0, 0xc00014a3c0, 0x5e, 0xa0, 0xc000162580, 0x50, 0x80)
         /home/jch/go/pkg/mod/github.com/pion/webrtc/v3@v3.0.0-20200716074551-070a7ea22eb9/internal/mux/endpoint.go:41 +0x59
 github.com/pion/srtp.(*SessionSRTCP).write(0xc0001150e0, 0xc000162580, 0x50, 0x80, 0x50, 0x80, 0x0)
         /home/jch/go/pkg/mod/github.com/pion/srtp@v1.4.0/session_srtcp.go:124 +0x152
 github.com/pion/srtp.(*WriteStreamSRTCP).Write(...)
         /home/jch/go/pkg/mod/github.com/pion/srtp@v1.4.0/stream_srtcp.go:132
 github.com/pion/webrtc/v3.(*PeerConnection).WriteRTCP(0xc00006a000, 0xc000458900, 0x2, 0x2, 0x2, 0xc000458900)
         /home/jch/go/pkg/mod/github.com/pion/webrtc/v3@v3.0.0-20200716074551-070a7ea22eb9/peerconnection.go:1451 +0xcb
 main.sendUpRTCP(0xc000021580, 0x0, 0x0)
         /home/jch/go/src/sfu/rtpconn.go:881 +0x71a
 main.rtcpUpSender(0xc000021580)
         /home/jch/go/src/sfu/rtpconn.go:887 +0x3c
 created by main.newUpConn.func1
         /home/jch/go/src/sfu/rtpconn.go:455 +0x53e
```

**Comment by jech** [comment]
I'm getting a mysterious crash in (*candidateBase).WriteTo.  To my eyes, it looks like `c.conn` is nil.
```
 panic: runtime error: invalid memory address or nil pointer dereference
 [signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x754bae]
 goroutine 237 [running]:
 github.com/pion/ice/v2.(*candidateBase).writeTo(0xc0000966e0, 0xc00014a3c0, 0x5e, 0xa0, 0xaeac20, 0xc000096790, 0x0, 0x0, 0x0)
         /home/jch/go/pkg/mod/github.com/pion/ice/v2@v2.0.0-rc.6/candidate_base.go:203 +0x4e
 github.com/pion/ice/v2.(*candidatePair).Write(...)
         /home/jch/go/pkg/mod/github.com/pion/ice/v2@v2.0.0-rc.6/candidatepair.go:86
 github.com/pion/ice/v2.(*Conn).Write(0xc00044b120, 0xc00014a3c0, 0x5e, 0xa0, 0xc00014a3c0, 0x5e, 0xa0)
         /home/jch/go/pkg/mod/github.com/pion/ice/v2@v2.0.0-rc.6/transport.go:105 +0x149
 github.com/pion/webrtc/v3/internal/mux.(*Endpoint).Write(0xc00018b5a0, 0xc00014a3c0, 0x5e, 0xa0, 0xc000162580, 0x50, 0x80)
         /home/jch/go/pkg/mod/github.com/pion/webrtc/v3@v3.0.0-20200716074551-070a7ea22eb9/internal/mux/endpoint.go:41 +0x59
 github.com/pion/srtp.(*SessionSRTCP).write(0xc0001150e0, 0xc000162580, 0x50, 0x80, 0x50, 0x80, 0x0)
         /home/jch/go/pkg/mod/github.com/pion/srtp@v1.4.0/session_srtcp.go:124 +0x152
 github.com/pion/srtp.(*WriteStreamSRTCP).Write(...)
         /home/jch/go/pkg/mod/github.com/pion/srtp@v1.4.0/stream_srtcp.go:132
 github.com/pion/webrtc/v3.(*PeerConnection).WriteRTCP(0xc00006a000, 0xc000458900, 0x2, 0x2, 0x2, 0xc000458900)
         /home/jch/go/pkg/mod/github.com/pion/webrtc/v3@v3.0.0-20200716074551-070a7ea22eb9/peerconnection.go:1451 +0xcb
 main.sendUpRTCP(0xc000021580, 0x0, 0x0)
         /home/jch/go/src/sfu/rtpconn.go:881 +0x71a
 main.rtcpUpSender(0xc000021580)
         /home/jch/go/src/sfu/rtpconn.go:887 +0x3c
 created by main.newUpConn.func1
         /home/jch/go/src/sfu/rtpconn.go:455 +0x53e
```

**Comment by Sean-Der** [comment]
@jech bleh :/

I bet a write is in flight, and the candidate was culled? (because of the restart)

The other thing it could be is from adding TCP-ICE Support.

Let me work on getting a reproduce! Reported here also https://github.com/pion/ice/issues/247

**Comment by jech** [comment]
> I bet a write is in flight, and the candidate was culled? (because of the
> restart)

In case it helps:

  - the browser was the offerer;
  - the PC had two tracks (one audio, one video), both in the
    browser->pion direction;
  - the restart was initiated by the browser (the JS called  pc.restartIce).

**Comment by jeremija** [comment]
@jech what does your `go.sum` say for `pion/ice` version?

**Comment by Sean-Der** [comment]
@jeremija I am 99% sure this is a ICE Restart only bug https://github.com/pion/ice/issues/247! So don't worry about debugging/looking :)

I will be able to easily fix tonight! Just going to add the null check and make `c.conn` an atomic, I don't want to take on more interlocking.

I think this started with 06922c1601abf902d8ed4fc9c0be6452ce216126 @jech to unblock yourself you could just use `v2.0.0-rc.3` for now! That is missing other bug fixes unfortunately :(

**Comment by jeremija** [comment]
Thanks! I also think it's highly unlikely to be related to ICE TCP since
the new commit hasn't been tagged yet.

**Comment by jech** [comment]
> @jech what does your go.sum say for pion/ice?

    github.com/pion/ice/v2 v2.0.0-rc.6 h1:Jz88W1iXzHBYJG6I5QbRTm+xuKD1vbgUW+NP1MtUPAg=
    github.com/pion/ice/v2 v2.0.0-rc.6/go.mod h1:VvpoDXwdierv9sPB8LAV3+T33ncCt0IG2NeI+CZYmTg=

Feel free to grab the source:

    git clone -b ice-restart https://www.irif.fr/~jch/software/sfu.git

**Comment by Sean-Der** [comment]
@jech I haven't confirmed this fixes the issue yet, but I have opened https://github.com/pion/ice/pull/248

I am pulling down `sfu` and will test now!

**Comment by jech** [comment]
This is now live on https://vps-63c87489.vps.ovh.net:8443/ .

Perhaps I'm doing something wrong, but I found it to be somewhat unreliable.  In particular, it doesn't seem to be able to renegotiate IP addresses smoothgly.  I'll keep debugging.

**Pull request merged**